File: Integrations/src/main/java/com/inverce/mod/integrations/support/recycler/MultiRecyclerAdapter.java
package com.inverce.mod.integrations.support.recycler;

import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.inverce.mod.core.functional.IFunction;
import com.inverce.mod.core.functional.IPredicate;
import com.inverce.mod.integrations.support.annotations.IBind;
import com.inverce.mod.integrations.support.annotations.IBinder;
import com.inverce.mod.integrations.support.annotations.ICreateVH;
import com.inverce.mod.v2.integrations.recycler.RecyclerAdapter;

import java.util.ArrayList;
import java.util.List;

public class MultiRecyclerAdapter<ITEM> extends RecyclerAdapter<ITEM, RecyclerView.ViewHolder> {
    protected List<MultiInfo<?, ?>> types;
    protected SparseArray<MultiInfo<?, ?>> typesSparse;

    public MultiRecyclerAdapter() {
        types = new ArrayList<>();
        typesSparse = new SparseArray<>();
    }

    @Override
    public final void onBindViewHolder(RecyclerView.ViewHolder holder, ITEM item, int position) {
        int type = getItemViewType(position);
        MultiInfo<?, ?> info = typesSparse.get(type);
        info.onBindViewHolder(holder, item, position);
    }

    @Override
    public final RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        MultiInfo<?, ?> info = typesSparse.get(viewType);
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        return info.createViewHolder.onCreateViewHolder(parent, inflater);
    }

    @Override
    public final int getItemViewType(int position) {
        ITEM item = getItem(position);
        for (MultiInfo<?, ?> type : types) {
            if (type.checkType.test(item)) {
                return type.registeredType;
            }
        }
        return -1;
    }

    public <I extends ITEM, VH extends RecyclerView.ViewHolder> int register(IPredicate<ITEM> checkType, IBind<I, VH> binder, ICreateVH<VH> createViewHolder) {
        MultiInfo<I, VH> info = new MultiInfo<>(checkType, binder, createViewHolder);
        types.add(info);
        info.registeredType = types.size();
        typesSparse.put(info.registeredType, info);
        return info.registeredType;
    }

    public <I extends ITEM, VH extends RecyclerView.ViewHolder> int register(IPredicate<ITEM> checkType, IBind<I, VH> binder, @NonNull IFunction<View, VH> createViewHolder, @LayoutRes int layout) {
        return register(checkType, binder, new ICreateVH<VH>() {
            @Override
            public VH onCreateViewHolder(ViewGroup parent, @NonNull LayoutInflater inflater) {
                return createViewHolder.apply(inflater.inflate(layout, parent, false));
            }
        });
    }

    public <I extends ITEM, VH extends RecyclerView.ViewHolder & IBinder<I>> int register(IPredicate<ITEM> checkType, ICreateVH<VH> createViewHolder) {
        return register(checkType, new IBind<I, VH>() {
            @Override
            public void onBindViewHolder(@NonNull VH vh, I item, int position) {
                vh.onBindViewHolder(item, position);
            }
        }, createViewHolder);
    }

    public <I extends ITEM, VH extends RecyclerView.ViewHolder & IBinder<I>> int register(IPredicate<ITEM> checkType, @NonNull IFunction<View, VH> createViewHolder, @LayoutRes int layout) {
        return register(checkType, new IBind<I, VH>() {
            @Override
            public void onBindViewHolder(@NonNull VH vh, I item, int position) {
                vh.onBindViewHolder(item, position);
            }
        }, new ICreateVH<VH>() {
            @Override
            public VH onCreateViewHolder(ViewGroup parent, @NonNull LayoutInflater inflater) {
                return createViewHolder.apply(inflater.inflate(layout, parent, false));
            }
        });
    }

    public <I extends ITEM> int register(IPredicate<ITEM> checkType, DataBinder<I> binder, @LayoutRes int layout) {
        return register(checkType, binder, BindViewHolder::new, layout);
    }

    protected class MultiInfo<I extends ITEM, VH extends RecyclerView.ViewHolder> {
        int registeredType;
        IPredicate<ITEM> checkType;
        IBind<I, VH> binder;
        ICreateVH<VH> createViewHolder;

        public MultiInfo(IPredicate<ITEM> checkType, IBind<I, VH> binder, ICreateVH<VH> createViewHolder) {
            this.checkType = checkType;
            this.binder = binder;
            this.createViewHolder = createViewHolder;
        }

        public void onBindViewHolder(RecyclerView.ViewHolder holder, ITEM item, int position) {
            //noinspection unchecked
            binder.onBindViewHolder((VH)holder, (I)item, position);
        }
    }

}

File: Integrations/src/main/java/com/inverce/mod/v2/integrations/recycler/MultiRecyclerAdapter.kt
package com.inverce.mod.v2.integrations.recycler

import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.ViewGroup
import java.util.*

open class MultiRecyclerAdapter<ITEM : Any> : RecyclerAdapter<ITEM, RecyclerView.ViewHolder>() {
    protected var types: MutableList<MultiInfo<*, *>> = ArrayList()
    protected var typesSparse: SparseArray<MultiInfo<*, *>> = SparseArray()

    fun resetRegistrations() {
        types = ArrayList()
        typesSparse = SparseArray()
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, item: ITEM, position: Int) {
        val type = getItemViewType(position)
        val info = typesSparse.get(type)
        info.onBindViewHolder(holder, item, position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val info = typesSparse.get(viewType)
        val inflater = LayoutInflater.from(parent.context)
        return info.createViewHolder(parent, inflater)
    }

    override fun getItemViewType(position: Int): Int {
        val item = getItem(position)
        for (type in types) {
            if (type.checkType(item)) {
                return type.registeredType
            }
        }
        return -1
    }

    inline fun <reified T> type(crossinline check: MultiPredicate<T> = { true }): MultiPredicate<ITEM> = { it is T && check(it as T) }

    infix fun <VH : RecyclerView.ViewHolder> MultiViewToVH<VH>.inflate(@LayoutRes layout: Int): MultiCreateVH<VH> = { parent: ViewGroup, inflater: LayoutInflater ->
        this(inflater.inflate(layout, parent, false))
    }

    @JvmOverloads
    fun <I : ITEM, VH : RecyclerView.ViewHolder> register(
            bind: MultiBind<I, VH>,
            vh: MultiCreateVH<VH>,
            check: MultiPredicate<ITEM> = { true }
    ) = MultiInfo(check, bind, vh).apply {
        registeredType = types.size
        types.add(this)
        typesSparse.put(registeredType, this)
    }.registeredType

    @JvmOverloads
    fun <I, VH> register(bind: MultiBind<I, VH>, create: MultiViewToVH<VH>, @LayoutRes res: Int, check: MultiPredicate<ITEM> = { true }): Int
            where I : ITEM, VH : RecyclerView.ViewHolder = register(bind, create inflate res, check)

    fun <I, VH> register(create: MultiViewToVH<VH>, @LayoutRes res: Int, check: MultiPredicate<ITEM> = { true }): Int
            where I : ITEM, VH : RecyclerView.ViewHolder, VH : MultiBinder<I> = register({ vh, item: I, pos ->
        vh.onBindViewHolder(item, pos)
    }, create inflate res, check)

    fun <I> register(binder: DataBinder<I>, @LayoutRes res: Int, check: MultiPredicate<ITEM> = { true }): Int
            where I : ITEM = register(binder::invoke, ::BindViewHolder inflate res, check)

    protected inner class MultiInfo<I : ITEM, VH : RecyclerView.ViewHolder>(
            var checkType: MultiPredicate<ITEM>,
            var binder: MultiBind<I, VH>,
            var createViewHolder: MultiCreateVH<VH>) {
        var registeredType: Int = 0
        fun onBindViewHolder(holder: RecyclerView.ViewHolder, item: ITEM, position: Int) {
            binder(holder as VH, item as I, position)
        }
    }

}